### PR TITLE
std.crypto: timing-safe functions to add&subtract serialized integers

### PR DESCRIPTION
**Pull request by jedisct1**
This is useful to increment nonces and for scalar reduction.

It will avoid code duplication in crypto/25519 and crypto/pcurves/*

### CODE REVIEW

File: lib/std/crypto/utils.zig [hunk 1]
```diff
@@ -77,6 +78,66 @@ pub fn timingSafeCompare(comptime T: type, a: []const T, b: []const T, endian: E
     return Order.lt;
 }
 
+/// Add two integers serialized as arrays of the same size, in constant time.
+/// The result is stored into `result`, and `true` is returned if the an overflow occurred.
```
Comment (daurnimator) [inline lib/std/crypto/utils.zig:82]: s/the an/any/

**Review by daurnimator — CHANGES_REQUESTED** [review]
Are there existing pieces of code that needs updating?

File: lib/std/crypto/utils.zig (hunk 1 above)
Comment (jedisct1) [reply lib/std/crypto/utils.zig]: Fixed, thanks!

**Comment by jedisct1** [comment]
The main reason I need these is a `reduce32()` function required for ECDSA. Other scalar reduction functions can then be updated to use `reduce32()` and additions.

File: lib/std/crypto/utils.zig [hunk 2]
```diff
@@ -77,6 +78,66 @@ pub fn timingSafeCompare(comptime T: type, a: []const T, b: []const T, endian: E
...
+/// Add two integers serialized as arrays of the same size, in constant time.
+/// The result is stored into `result`, and `true` is returned if an overflow occurred.
+pub fn timingSafeAdd(comptime T: type, a: []const T, b: []const T, result: []T, endian: Endian) bool {
+    const len = a.len;
+    debug.assert(len == b.len and len == result.len);
+    const bits = switch (@typeInfo(T)) {
+        .Int => |cinfo| if (cinfo.signedness != .unsigned) @compileError("Elements to be added must be unsigned") else cinfo.bits,
+        else => @compileError("Elements to be added must be integers"),
+    };
+    const Cext = std.meta.Int(.unsigned, bits + 1);
```
Comment (LemonBoy) [inline lib/std/crypto/utils.zig:90]: The `+1` trick is incredibly wasteful as it forces the sum to work on 33/65 bits. Use `@addWithOverflow` as it's done in the bigint implementation instead.

Comment (jedisct1) [reply lib/std/crypto/utils.zig:90]: Is @addWithOverflow guaranteed to be constant-time?

The intent of using a bigger integer is to minimize the chances of having conditional branches.

Comment (LemonBoy) [reply lib/std/crypto/utils.zig:90]: > Is @addWithOverflow guaranteed to be constant-time?

No, but neither is the"bigint" arithmetic that LLVM legalizes this operation to. If the target CPU has a carry flag and a way of transferring it to a GPR, a pretty safe assumption to make, `@addWithOverflow` is cheap and constant-time.

Comment (jedisct1) [reply lib/std/crypto/utils.zig]: The code has been changed to use `@addWithOverflow` and `@subWithOverflow`.

Thanks!

**Pull request merged**